File: 0986-interval-list-intersections/0986-interval-list-intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList,
                                             vector<vector<int>>& secondList) {
        int i = 0;
        int j = 0;
        vector<vector<int>> res;

        while (i < firstList.size() && j < secondList.size()) {
            int start = max(firstList[i][0], secondList[j][0]);
            int end = min(firstList[i][1], secondList[j][1]);

            if (start <= end) {
                res.push_back({start, end});
            }

            int end1 = firstList[i][1];
            int end2 = secondList[j][1];
            if (end1 < end2) {
                i++;
            } else {
                j++;
            }
        }
        return res;
    }
};

File: 0986-interval-list-intersections/0986-interval-list-intersections.go
func intervalIntersection(firstList [][]int, secondList [][]int) [][]int {
	var i int = 0
	var j int = 0
	res := [][]int{}

	for i < len(firstList) && j < len(secondList) {
		start := max(firstList[i][0], secondList[j][0])
		end := min(firstList[i][1], secondList[j][1])

		if start <= end {
			res = append(res, []int{start, end})
		}

		end1 := firstList[i][1]
		end2 := secondList[j][1]

		if end1 < end2 {
			i++
		} else {
			j++
		}
	}
	return res
}